src/server/tablet_snap.rs

```diff
@@ -0,0 +1,526 @@
+        let key = TabletSnapKey::from_region_snap(
+            meta.get_region_id(),
+            meta.get_to_peer().get_id(),
+            meta.get_message().get_snapshot(),
+        );
+
+        let data = snapshot.get_data();
+        let mut snapshot_data = RaftSnapshotData::default();
+        snapshot_data.merge_from_bytes(data)?;
+        let snapshot_meta = snapshot_data.take_meta();
+        let io_type = if snapshot_meta.get_for_balance() {
+            IoType::LoadBalance
+        } else {
+            IoType::Replication
+        };
+        let _with_io_type = WithIoType::new(io_type);
+
+        Ok(RecvTabletSnapContext {
+            key,
+            raft_msg: meta,
+            io_type,
+            start: Instant::now(),
+        })
+    }
+
+    fn finish<R: RaftStoreRouter<impl KvEngine>>(self, raft_router: R) -> Result<()> {
+        let _with_io_type = WithIoType::new(self.io_type);
+        let key = self.key;
+        if let Err(e) = raft_router.send_raft_msg(self.raft_msg) {
+            return Err(box_err!("{} failed to send snapshot to raft: {}", key, e));
+        }
+        info!("saving all snapshot files"; "snap_key" => %key, "takes" => ?self.start.saturating_elapsed());
+        Ok(())
+    }
+}
+
+async fn send_snap_files(
+    mgr: &TabletSnapManager,
+    mut sender: impl Sink<(SnapshotChunk, WriteFlags), Error = Error> + Unpin,
+    msg: RaftMessage,
+    key: TabletSnapKey,
+    limiter: Limiter,
+) -> Result<u64> {
+    let path = mgr.get_tablet_checkpointer_path(&key);
+    info!("begin to send snapshot file";"snap_key" => %key);
+    let files = fs::read_dir(&path)?
+        .map(|d| Ok(d?.path()))
+        .collect::<Result<Vec<_>>>()?;
+    let mut total_sent = msg.compute_size() as u64;
+    let mut chunk = SnapshotChunk::default();
+    chunk.set_message(msg);
+    sender
+        .feed((chunk, WriteFlags::default().buffer_hint(true)))
+        .await?;
+    for path in files {
+        let name = path.file_name().unwrap().to_str().unwrap();
+        let mut buffer = Vec::with_capacity(SNAP_CHUNK_LEN);
+        buffer.push(name.len() as u8);
+        buffer.extend_from_slice(name.as_bytes());
+        let mut f = File::open(&path)?;
+        let mut off = buffer.len();
+        loop {
+            unsafe { buffer.set_len(SNAP_CHUNK_LEN) }
+            let readed = f.read(&mut buffer[off..])?;
+            limiter.consume(readed);
+            let new_len = readed + off;
+            total_sent += new_len as u64;
+            unsafe {
+                buffer.set_len(new_len);
+            }
+            let mut chunk = SnapshotChunk::default();
+            chunk.set_data(buffer);
+            sender
+                .feed((chunk, WriteFlags::default().buffer_hint(true)))
+                .await?;
+            if new_len < SNAP_CHUNK_LEN {
+                break;
+            }
+            buffer = Vec::with_capacity(SNAP_CHUNK_LEN);
+            off = 0
```

Comment:
the second SnapshotChunk will not have the file name?  But the receiver side will not know it and it will try to parse the file name anyway which will fail. I think here the buf should always include the file name

Reply:
yes，the second snapshot chunk should not contain the file name, the file name only exists in the first snapshot chunk. We can use the snapshot chunk's data len to check the file has sent all . if the length < 1M , it means that the file has sent all. BTW, if the data's len is 1M exactly, it will send one snapshot chunk, but not carry any data .